### DISCUSSION TITLE
React Vite remote + Webpack Host

### DISCUSSION THREAD

**Discussion by colindunn109** [originjs/vite-plugin-federation#421] [General]
I saw that in the README, there is a line:
>It is not recommended to mix Vite and Webpack in React projects, as there is no guarantee that Vite/Rollup and Webpack will generate the same chunk when packaging commonjs, which may cause problems with shared.

I would love to know more about this. What is actually happening under the hood that causes this issue with React apps in particular? Is there any way that we can add config to make the packaging process more consistent?

handshake between Vite and Webpack is a major usecase for our app and so being able to unlock this functionality without changing the Host app (Webpack) would be ideal. 

Thank you!

**Comment by Jiannan-dev** [comment]
The detailed reason is that I can't guarantee that multiple packaging tools have consistent exports, so you can understand that vite in dev mode will use esbuild packaging library, serve mode will use rollup, webpack are using their own written code packaging
For ESM (e.g. vue), the packaging tool always ensures that the export is consistent
However, for non-ESM compliant libraries (e.g. React), there is currently no guarantee that multiple packaging tools will have consistent exports, and different packaging tools will convert CJS (CommonJS) exports differently, which makes it impossible to share the shared because there is no identical export structure

**Comment by Jiannan-dev** [comment]
> It is not recommended to mix Vite and Webpack in React projects, as there is no guarantee that Vite/Rollup and Webpack will generate the same chunk when packaging commonjs, which may cause problems with shared.

To be precise, this sentence is incomplete, not only react, as long as it is the shared library of the CJS specification, there will be this problem.

**Comment by Jiannan-dev** [comment]
https://github.com/originjs/vite-plugin-federation/issues/413#issuecomment-1543766956
Maybe we can see if this approach solves part of the problem

**Comment by colindunn109** [comment]
@ruleeeer Super insightful and as I dug more thats kinda what I was finding, though I was having a hard time reasoning why it was React only problem.

 Seems like if you add some heavy config to Webpack to enable ESM to be consumed this can be worked around. This [demo video](https://youtu.be/t-nchkL9yIg?t=1199 )is pretty helpful on seeing it come together.
 
 Though this feels pretty hacky anyway.

After chatting through it with my team we basically landed on just keep the tooling the same on host and remote. Trying to do a bunch of fancy stuff to get differing tooling to agree on modules just leads to more of a headache.

Seems like you've arrived at the same position. Appreciate the response 👍
